### ISSUE TITLE
Multiple Gets based on Return values in API return value

### ISSUE THREAD

**Issue by pmaldonato** [geral2/SQL-APIConsumer#34]
I've been banging my head against a wall for a week. I need to loop through this API call (and others) based on the values of the first 3 values, I' can pull 20 records at a time but I don't know where to put my loop, do I need a cursor? This is abbreviated, I've already pulled and parsed my data, but can never get to loop to the next 20.:
--Call EROAD Web Service API
--Set offsets and local variables

DECLARE @counter  integer = -1
DECLARE @total	integer = 0
DECLARE @offset integer = 0

--Access RESTFUL API

--Set Header with ApiKey must be in the header

DECLARE @header nvarchar(max) = 
  '[{
		"Name": "Content-Type",
		"Value" :"application/json; charset=utf-8"
	},
	{
		"Name": "X-Host",
		"Value" :"https://api.na.eroad.com/v1"
	},{
		"Name": "ApiKey",
		"Value" :"xxXZNL1bx4hpeAEmTnpszhuRTq7xCtFj" -- not actual ApiKey
	}]';
--Set URL for login

DECLARE @wurl varchar(max) = 'https://api.na.eroad.com/v1/drivers?limit=20&offset=' + convert(nvarchar(5),@offset)

WHILE (@total <> @offset + @counter) 

BEGIN 

DECLARE @ts as table
(
	Json_Result nvarchar(max)
	,ContentType varchar(100),
	ServerName varchar(100),
	Statuscode varchar(100),
	[Description] varchar(100),
	Json_Headers nvarchar(max)
)

DECLARE @i as int
INSERT INTO @ts
--Get Results
 exec @i = [dbo].[APICaller_GET_Extended] -- need for apikey
 							@wurl
							,''
							,@header
 
--set variables to json record value @ts.json_result starts with {"count":20,"offset":0,"total":396,"drivers":["
SET @counter = ( SELECT * FROM OPENJSON((select Json_Result from @ts)) WITH ( "count" integer '$."count"'))   
SET @total = ( SELECT * FROM OPENJSON((select Json_Result from @ts)) WITH ( "total"	integer '$."total"'))	  		
SET @offset = ( SELECT *  FROM OPENJSON((select Json_Result from @ts)) WITH ( "offset" integer '$."offset"'))

INSERT INTO [dbo].[jdrivers]
           ([BulkColumn])
		   VALUES((select Json_result from @ts ))

SELECT @offset = (@offset + 20)

END
----------

I need to loop through the results until I I reach the count. Authentication is a header only. So I tried to use APICaller_Get after APICaller_Get_Extended and it failed authentication on the rows.. Here's my fully working with SQL Statement:

 --Set Header
Declare @header nvarchar(max) = 
  '[{
		"Name": "Content-Type",
		"Value" :"application/json; charset=utf-8"
	},
	{
		"Name": "X-Host",
		"Value" :"https://api.na.eroad.com/v1"
	},{
		"Name": "ApiKey",
		"Value" :"spXZNL0bxxxxxxxxpszhuRTq7xCtFj"
	}]';
--Set URL
Declare @wurl varchar(max) = 'https://api.na.eroad.com/v1/drivers?limit=20&offset=0'

Declare @ts as table
(
	Json_Result nvarchar(max)
	,ContentType varchar(100),
	ServerName varchar(100),
	Statuscode varchar(100),
	[Description] varchar(100),
	Json_Headers nvarchar(max)
)

Declare @i as int 
 
INSERT INTO @ts
 --Get Driver Data
 exec @i = [dbo].[APICaller_GET_Extended]
 							@wurl
							,''
							,@header

SELECT * FROM @ts
 
SELECT  * 
 FROM OPENJSON((select Json_Headers from @ts))  
		WITH (   
				 Header		nvarchar(max) '$."Name"'      
				,Value		nvarchar(max) '$."Value"'      
				) a

SELECT *
 FROM OPENJSON((select Json_Result from @ts))  
		WITH (   
				 "count"		integer '$."count"'      
				,offset			integer '$."offset"'      
				,total			integer '$."total"' 
				) a

DECLARE @drivers NVARCHAR(MAX)

SET @drivers= ( 
SELECT drivers
 FROM OPENJSON((select Json_Result from @ts))  
		WITH (  
				drivers	nvarchar(max)   as json  
				) )

   
SELECT * FROM  
 OPENJSON ( @drivers ) 
WITH (   
              id_driver		    nvarchar(max)'$.id' ,  
              firstname			nvarchar(max)'$.firstName',  
              lastname			nvarchar(max)'$.lastName',  
              pin				nvarchar(max)'$.pin',
			  id_shift			nvarchar(max)'$.groups[0].id' ,
			  sfift_name		nvarchar(max)'$.groups[0].name',
			  lic_type			nvarchar(max)'$.licenses[1].type',
			  lic_number		nvarchar(max)'$.licenses[1].number',
			  lic_juris			nvarchar(max)'$.licenses[1].jurisdiction',
			  lic_Prinary		nvarchar(max)'$.licenses[1].isPrimary',
			  expiration		nvarchar(max)'$.licenses[1].expiration',
			  classification	nvarchar(max)'$.licenses[1].classification',
			  endorsements		nvarchar(max)'$.licenses[1].endorsements',
			  restrictions		nvarchar(max)'$.licenses[1].restrictions',
			  active			nvarchar(max)'$.active',
			  hoursOfService	nvarchar(max)'$.hoursOfService',
			  terminal			nvarchar(max)'$.terminal',
			  driverExternalId	nvarchar(max)'$.driverExternalId'   
			 
 ) 

---------------------------

Full Schema

{
  "count": 0,
  "offset": 0,
  "total": 0,
  "drivers": [
    {
      "id": "string",
      "firstName": "string",
      "lastName": "string",
      "alias": "string",
      "email": "string",
      "mobileNumber": "string",
      "pin": "string",
      "groups": [
        {
          "id": "string",
          "name": "string"
        }
      ],
      "licenses": [
        {
          "type": "string",
          "number": "string",
          "jurisdiction": "string",
          "isPrimary": true,
          "expiration": "2020-06-27",
          "classification": "string",
          "endorsements": "string",
          "restrictions": "string"
        }
      ],
      "active": true,
      "hoursOfService": "string",
      "terminal": "string",
      "driverExternalId": "string"
    }
  ]
}

How to pass back the off set into the @url and loop back through has eluded me

**Comment by geral2** [comment]
Hi @pmaldonato,

This does'nt look that hard, but I would need more info... For example what should be the next execution?

https://api.na.eroad.com/v1/drivers?limit=20&offset=1

?

**Comment by pmaldonato** [comment]
Hey Gerald,


It would be

https://api.na.eroad.com/v1/drivers?limit=20&offset=20

On this one, I’m trying to pull all drivers and the max record pull at a time is 20 for this endpoint.

Thanks fo the help.
Paul

**Comment by geral2** [comment]
I don't think you really need the extended method, but could you try with the code below?

```
--Call EROAD Web Service API
--Set offsets and local variables

DECLARE @counter integer = -1
DECLARE @ToTal integer = 0
DECLARE @offset integer = 0
DECLARE @acc integer = 0

--Access RESTFUL API

--Set Header with ApiKey must be in the header

DECLARE @Header nvarchar(max) =
'[{
"Name": "Content-Type",
"Value" :"application/json; charset=utf-8"
},
{
"Name": "X-Host",
"Value" :"https://api.na.eroad.com/v1"
},{
"Name": "ApiKey",
"Value" :"xxXZNL1bx4hpeAEmTnpszhuRTq7xCtFj" -- not actual ApiKey
}]';
--Set URL for login

DECLARE @wurl varchar(max) = 'https://api.na.eroad.com/v1/drivers?limit=20&offset=' + convert(nvarchar(5),@offset)

DECLARE @ts as table
(
Json_Result nvarchar(max)
,ContentType varchar(100),
ServerName varchar(100),
Statuscode varchar(100),
[Description] varchar(100),
Json_Headers nvarchar(max)
)

DECLARE @i as int

WHILE (@ToTal < @offset ) 
BEGIN

	BEGIN TRY
		SET @wurl = 'https://api.na.eroad.com/v1/drivers?limit=20&offset=' + convert(nvarchar(5),@offset)

		INSERT INTO @ts
		--Get Results
		exec @i = [dbo].[APICaller_GET_Extended] -- need for apikey
						@wurl
						,''
						,@Header

		--set variables to json record value @ts.json_result starts with {"count":20,"offset":0,"total":396,"drivers":["

			SELECT @counter = A.[count]
				  ,@ToTal	= A.Total
				  ,@offset	= a.offset
			 FROM OPENJSON((select Json_Result from @ts))
			 WITH (
					"count" INT '$."count"'
					,offset INT '$."offset"'
					,total  INT '$."total"'
				  ) a

		--SET @counter = ( SELECT * FROM OPENJSON((select Json_Result from @ts)) WITH ( "count" integer '$."count"'))
		--SET @ToTal = ( SELECT * FROM OPENJSON((select Json_Result from @ts)) WITH ( "total" integer '$."total"'))
		--SET @offset = ( SELECT * FROM OPENJSON((select Json_Result from @ts)) WITH ( "offset" integer '$."offset"'))

		INSERT INTO [dbo].[jdrivers]
		([BulkColumn])
		VALUES((select Json_result from @ts ))

	 END TRY
	 BEGIN CATCH
		THROW;
		BREAK
	 END CATCH

	 PRINT @offset

	 SELECT @offset = (@offset + 20)
	 SET @acc = @acc + 1;

END -- While
 
```

**Comment by pmaldonato** [comment]
Gerald,

This is to access an API developed using  https://swagger.io/specification/v2/ API by EROAD Fleet Management.

This API only allows the ApiKey to be in the header. Depending on who designed the API, it can be in the @url or the @header. This one is @header only, so I had to use _Extended. I got it to work!

Thanks for all your help. I’ve attached a script if you want to through it in there for a swagger API  example. For me this finally looped through 21 rows containing  396 records 20 at a time until all are read.  The last record was

BulkColumn
{"count":0,"offset":396,"total":396,"drivers":[]}

Thanks again!

Paul Maldonato
Emerge Business Consulting
(415) 305-8758

paul@emergebc.com<mailto:paul@emergebc.com>
www.emergebc.com<http://www.emergebc.com/>

Click Here for Support<https://www.gotomeet.me/PaulMaldonato>
